kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/storage/PartitionLog.java

```diff
@@ -275,53 +262,136 @@ private void publishMessages(final Optional<PersistentTopic> persistentTopicOpt,
+        final int numMessages = encodeResult.getNumMessages();
+        final ByteBuf byteBuf = encodeResult.getEncodedByteBuf();
         final long beforePublish = time.nanoseconds();
-        persistentTopic.publishMessage(byteBuf,
-                MessagePublishContext.get(offsetFuture, persistentTopic, numMessages, System.nanoTime()));
+
+        CompletableFuture<Long> offsetFuture;
+
+        // For control message we don't need to check deduplication.
+        if (appendInfo.isControlBatch()) {
+            offsetFuture = publishControlMessage(persistentTopic, byteBuf, numMessages);
+        } else {
+            offsetFuture = publishNormalMessage(persistentTopic, byteBuf, appendInfo);
+        }
+
         offsetFuture.whenComplete((offset, e) -> {
             appendRecordsContext.getCompleteSendOperationForThrottling().accept(byteBuf.readableBytes());
-            encodeResult.recycle();
+
             if (e == null) {
                 requestStats.getMessagePublishStats().registerSuccessfulEvent(
                         time.nanoseconds() - beforePublish, TimeUnit.NANOSECONDS);
                 final long lastOffset = offset + numMessages - 1;
+
+                AnalyzeResult analyzeResult = analyzeAndValidateProducerState(
+                        encodeResult.getRecords(), Optional.of(offset), AppendOrigin.Client);
                 analyzeResult.updatedProducers().forEach((pid, producerAppendInfo) -> {
                     if (log.isDebugEnabled()) {
                         log.debug("Append pid: [{}], appendInfo: [{}], lastOffset: [{}]",
                                 pid, producerAppendInfo, lastOffset);
                     }
-                    // When we have real start offset, update current txn first offset.
-                    producerAppendInfo.updateCurrentTxnFirstOffset(appendInfo.isTransaction(), offset);
                     producerStateManager.update(producerAppendInfo);
                 });
                 analyzeResult.completedTxns().forEach(completedTxn -> {
                     // update to real last offset
-                    completedTxn.lastOffset(lastOffset);
+                    completedTxn.lastOffset(lastOffset - 1);
                     long lastStableOffset = producerStateManager.lastStableOffset(completedTxn);
                     producerStateManager.updateTxnIndex(completedTxn, lastStableOffset);
                     producerStateManager.completeTxn(completedTxn);
                 });
+
                 appendFuture.complete(offset);
             } else {
                 log.error("publishMessages for topic partition: {} failed when write.", fullPartitionName, e);
                 requestStats.getMessagePublishStats().registerFailedEvent(
                         time.nanoseconds() - beforePublish, TimeUnit.NANOSECONDS);
-                appendFuture.completeExceptionally(Errors.KAFKA_STORAGE_ERROR.exception());
```

Comment:
@BewareMyPower See here, all exceptions will be treated as `Errors.KAFKA_STORAGE_ERROR`, I don't want to change previous behavior.